xtevent/_eventgenvars.ado

```diff
@@ -1,306 +1,500 @@
+		cap gen `notmiss'=!missing(`z')
+		
+		cap by `panelvar' (`timevar'): gen `zt'=`timevar' if `notmiss'==1 
+		cap by `panelvar' (`timevar'): egen `maxzt'=max(`zt') 
+		cap by `panelvar' (`timevar'): egen `minzt'=min(`zt')
+
+		*first filter: all units satisfy the bounds condition? 
+		if `bin'==1 & `norever'==1 {
+			*verify the lower-bound value
+			cap assert `z'==`rminz' if `minzt'==`timevar' & `touse'
+			* if the lower-bound value is zero, then test the upper-bound
+			if !_rc{
+				cap assert `z'==`rmaxz' if `maxzt'==`timevar' & `touse'
+				if !_rc loc bounds 1
+			} 
+		}
+		
+		*second filter: if some units didn't satisfy the first filter, then allow those units to have only adopted policy or only unadopted policy (and no missing values inside the observed range)
+		if `bin'==1 & `norever'==1 & `bounds'==0 {
+		
+		tempvar ilb iub sb sbmin
+		cap by `panelvar' (`timevar'): gen `ilb'=(`z'==`rminz') if `minzt'==`timevar' & `touse'
+		cap by `panelvar' (`timevar'): gen `iub'=(`z'==`rmaxz') if `maxzt'==`timevar' & `touse'
+		cap egen `sb'=rowtotal(`ilb' `iub') if (`minzt'==`timevar' | `maxzt'==`timevar')
+		*sbmin is an indicator of the units that satisfied the first filter 
+		cap by `panelvar' (`timevar'): egen `sbmin'=min(`sb') if `timevar'>=`zmint2' & `timevar'<=`zmaxt2' & `touse'
+		
+		cap assert inlist(`z',`rminz',`rmaxz') if `sbmin'==0
+		if !_rc loc bounds 1
+		}
+	
+	
+		if `bounds'==0 & "`impute'"=="stag" | "`impute'"=="instag" {
+				di "For some units, the changes in policyvar are not consistent with no-unobserved-change. Reverting to default (no imputation)."
+				loc impute =""	
+		}
+	}
+	
+	***************** no unobserved change ***************************
+
+	if "`impute'"!="" {
+		cap replace zn2=`zminv2' if `timevar'<`zmint2'
+		cap replace zn2=`zmaxv2' if `timevar'>`zmaxt2'
+	}
+
+	************** impute inner missing values ***********************
+	if "`impute'"=="instag" {
+		
+		tempvar zdown zup 
+		cap gen `zdown'=`z'
+		gsort `panelvar' `timevar'
+		cap replace `zdown'=`zdown'[_n-1] if missing(`zdown')
+		cap gen `zup'=`z'
+		gsort + `panelvar' - `timevar'
+		cap replace `zup'=`zup'[_n-1] if missing(`zup')
+		sort `panelvar' `timevar'
+		
+		cap replace zn2=`zdown' if `timevar'>=`zmint2' & `timevar'<=`zmaxt2' & missing(`z') & `zdown'==`zup'
+		
+	}
+
+****************************** event-time dummies ***********************
+	if "`static'"==""{
+		qui xtset `panelvar' `timevar', noquery
+		
+		qui sort `panelvar' `timevar', stable
+			
+		* Generate event time dummies 
+		* First I generate all possible event time dummies to have event time for the trend regressions, then I keep only the ones in the window.
+		* While inefficient, this allows me to have event-time outside the window 
+		* Fix: This is too inefficient for large datasets. Only generate in window
+		* quietly forv klevel=-`tdiff'(1)`tdiff' {
+		
+		*create z delta
+		tempvar zd 
+		qui gen `zd'=zn2- L1.zn2
+		
+		*matrix limits

```

Comment:
@Constantino-Carreto-Romero It's unclear what this comment is for. Maybe remove it?

Reply:
I changed it for a clearer explanation